src/serac/physics/utilities/functional/boundary_integral.hpp

```diff
@@ -5,215 +5,40 @@
-}
-/**
- * @overload
- * @note Intended to be used with Serac's finite element space types
- */
-template <typename space, typename T>
-auto Load(const T& u, int e)
-{
-  if constexpr (space::components == 1) {
-    return detail::Load<space::ndof>(u, e);
-  }
-  if constexpr (space::components > 1) {
-    return detail::Load<space::ndof, space::components>(u, e);
-  }
-};
-
-/**
- * @brief Adds the contributions of the local residual to the global residual
- * @param[inout] r_global The full element-decomposed residual
- * @param[in] r_local The contributions to a residual from a single element
- * @param[in] e The index of the element whose residual is @a r_local
- */
-template <int ndof>
-void Add(const mfem::DeviceTensor<2, double>& r_global, tensor<double, ndof> r_local, int e)
-{
-  for (int i = 0; i < ndof; i++) {
-    r_global(i, e) += r_local[i];
-  }
-}
-/**
- * @overload
- * @note Used when each node has multiple DOFs
- */
-template <int ndof, int components>
-void Add(const mfem::DeviceTensor<3, double>& r_global, tensor<double, ndof, components> r_local, int e)
-{
-  for (int i = 0; i < ndof; i++) {
-    for (int j = 0; j < components; j++) {
-      r_global(i, j, e) += r_local[i][j];
-    }
-  }
-}
-
-/**
- * @brief Computes the arguments to be passed into the q-function (shape function evaluations)
- * By default:
- *  H1 family elements will compute {value, gradient}
- *  Hcurl family elements will compute {value, curl}
- *  TODO: Hdiv family elements will compute {value, divergence}
- *  TODO: L2 family elements will compute value
- *
- * In the future, the user will be able to override these defaults
- * to omit unused components (e.g. specify that they only need the gradient)
- *
- * @param[in] u The DOF values for the element
- * @param[in] xi The position of the quadrature point in reference space
- * @param[in] J The Jacobian of the element transformation at the quadrature point
- * @tparam element_type The type of the element (used to determine the family)
- */
-template <typename element_type, typename T, int dim>
-auto Preprocess(T u, const tensor<double, dim> xi, const tensor<double, dim, dim> J)
-{
-  // TODO: Static assert that element_type is an element??
-  if constexpr (element_type::family == Family::H1) {
-    return std::tuple{dot(u, element_type::shape_functions(xi)),
-                      dot(u, dot(element_type::shape_function_gradients(xi), inv(J)))};
-  }
-
-  if constexpr (element_type::family == Family::HCURL) {
-    // HCURL shape functions undergo a covariant Piola transformation when going
-    // from parent element to physical element
-    auto value = dot(u, dot(element_type::shape_functions(xi), inv(J)));
-    auto curl  = dot(u, element_type::shape_function_curl(xi) / det(J));
-    if constexpr (dim == 3) {
-      curl = dot(curl, transpose(J));
-    }
-    return std::tuple{value, curl};
-  }
-}
+namespace boundary_integral {
```

Comment:
My intention was to put the things shared by `integral.hpp` (which should probably be renamed to `domain_integral.hpp`) and `boundary_integral.hpp` in `integral_utilities.hpp`, and put the things that are different in their individual files, behind a unique namespace. If they're all in the `detail` namespace, some of the `domain_integral`-specific definitions might conflict with the `boundary_integral`-specific ones. Right now, there should not a problem because those functions (Preprocess, Postprocess, etc) actually have different function signatures (so they could just be an overload set), as a result of our current inability to get tangential derivative info from mfem.

Reply:
Yeah, I'm ok with either the overload set or namespace solution as long as the domain and boundary integral functions have the same layout.